### PR TITLE
Tests ulyanov 2016 modules

### PR DESCRIPTION
**Pull request by jbueltemeier**

### CODE REVIEW

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 1]
```diff
@@ -76,6 +109,20 @@ def test_ulyanov_et_al_2016_upsample(subtests):
         assert module.mode == "nearest"
 
 
+def test_UlyanovEtAl2016HourGlassBlock(subtests):
+    intermediate = nn.Conv2d(3, 3, 1)
+    sequential = modules.UlyanovEtAl2016HourGlassBlock(intermediate)
+
+    assert isinstance(sequential, modules.SequentialWithOutChannels)
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:116]:
```suggestion
    assert isinstance(sequential, modules.UlaynovEtAl2016HourGlassBlock)
```

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 2]
```diff
@@ -76,6 +109,20 @@ def test_ulyanov_et_al_2016_upsample(subtests):
         assert module.mode == "nearest"
 
 
+def test_UlyanovEtAl2016HourGlassBlock(subtests):
+    intermediate = nn.Conv2d(3, 3, 1)
+    sequential = modules.UlyanovEtAl2016HourGlassBlock(intermediate)
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:114]: Why `sequential`? Wouldn't `hour_glass` or `hour_glass_block` be a better fit here?

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 3]
```diff
@@ -76,6 +109,20 @@ def test_ulyanov_et_al_2016_upsample(subtests):
...
+
+    assert isinstance(sequential, modules.SequentialWithOutChannels)
+
+    with subtests.test("modules"):
+        assert len(sequential) == 3
+        sequential_modules = tuple(module for module in sequential.children())
+        assert isinstance(sequential_modules[0], nn.Module)
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:121]:
```suggestion
        assert isinstance(sequential.down, modules.UlyanovEtAl2016StylizationDownsample)
```

Same for the other two. You could wrap them into individual subtests.

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 4]
```diff
@@ -120,3 +167,187 @@ def test_get_activation_module(subtests):
...
+        with subtests.test("channel_dim"):
+            assert block.channel_dim == channel_dim
+
+        with subtests.test("forward"):
+            inputs = (input_image,) * 2
+            output_image = block(*inputs)
+            assert isinstance(output_image, torch.Tensor)
+
+
+def test_UlyanovEtAl2016JoinBlock_wrong_input():
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:257]:
```suggestion
def test_UlyanovEtAl2016JoinBlock_num_channels_names_mismatch():
```

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 5]
```diff
@@ -120,3 +167,187 @@ def test_get_activation_module(subtests):
...
+        block = modules.UlyanovEtAl2016JoinBlock(
+            branch_in_channels, names=names, channel_dim=channel_dim
+        )
+
+        with subtests.test("names"):
+            assert (
+                tuple(block._modules.keys()) == names
+                if names is not None
+                else ("0", "1")
+            )
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:241]: This is irrelevant. You should check if `block.block1` and so on are actually norm modules.

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 6]
```diff
@@ -120,3 +167,187 @@ def test_get_activation_module(subtests):
...
+        with subtests.test("out_channels"):
+            assert block.out_channels == sum(branch_in_channels)
+
+        with subtests.test("channel_dim"):
+            assert block.channel_dim == channel_dim
+
+        with subtests.test("forward"):
+            inputs = (input_image,) * 2
+            output_image = block(*inputs)
+            assert isinstance(output_image, torch.Tensor)
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:254]: You need to actually check the output for example with [`instance_norm`](https://pytorch.org/docs/stable/nn.functional.html?highlight=instance_norm#torch.nn.functional.instance_norm).

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 7]
```diff
@@ -120,3 +167,187 @@ def test_get_activation_module(subtests):
...
+        assert isinstance(block.join, modules.UlyanovEtAl2016JoinBlock)
+
+    with subtests.test("out_channels"):
+        assert (
+            block.out_channels == deep_branch.out_channels + shallow_branch.out_channels
+        )
+
+    with subtests.test("forward"):
+        output_image = block(input_image)
+        assert isinstance(output_image, torch.Tensor)
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:284]: Same as above. Please check the actual output.

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 8]
```diff
@@ -120,3 +167,187 @@ def test_get_activation_module(subtests):
...
+                assert isinstance(
+                    transformer[1],
+                    nn.Conv2d if impl_params else modules.UlyanovEtAl2016ConvBlock,
+                )
+                output_conv = transformer[1] if impl_params else transformer[1][0]
+                assert output_conv.out_channels == 3
+                assert output_conv.kernel_size == to_2d_arg(1)
+                assert output_conv.stride == to_2d_arg(1)
+
+            with subtests.test("forward"):
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:346]:
```suggestion
            with subtests.test("forward size"):
```

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 9]
```diff
@@ -76,6 +109,20 @@ def test_ulyanov_et_al_2016_upsample(subtests):
         assert module.mode == "nearest"
 
 
+def test_UlyanovEtAl2016HourGlassBlock(subtests):
+    intermediate = nn.Conv2d(3, 3, 1)
+    hour_glass = modules.UlyanovEtAl2016HourGlassBlock(intermediate)
+
+    assert isinstance(hour_glass, modules.UlyanovEtAl2016HourGlassBlock)
+
+    with subtests.test("Downsample"):
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:118]:
```suggestion
    with subtests.test("down"):
```

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 10]
```diff
@@ -76,6 +109,20 @@ def test_ulyanov_et_al_2016_upsample(subtests):
...
+        assert isinstance(hour_glass.down, modules.UlyanovEtAl2016StylizationDownsample)
+    with subtests.test("intermediate"):
+        assert isinstance(hour_glass.intermediate, type(intermediate))
+    with subtests.test("Upsample"):
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:122]:
```suggestion
    with subtests.test("up"):
```

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 11]
```diff
@@ -120,3 +167,217 @@ def test_get_activation_module(subtests):
...
+    padding = (1, 1, 1, 1)
+    conv_block = modules.UlyanovEtAl2016ConvBlock(
+        in_channels, out_channels, kernel_size, stride=stride
+    )
+
+    assert isinstance(type(conv_block), type(modules.SequentialWithOutChannels))
+
+    with subtests.test("modules"):
+        assert len(conv_block) == 4
+        assert isinstance(type(conv_block[0]), type(nn.ReflectionPad2d))
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:185]: See discussion in https://github.com/pmeier/pystiche_papers/pull/82#pullrequestreview-458947475

```suggestion
        assert isinstance(conv_block[0], nn.ReflectionPad2d)
```

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 12]
```diff
@@ -120,3 +167,217 @@ def test_get_activation_module(subtests):
...
+def test_UlyanovEtAl2016JoinBlock(subtests, input_image):
+    in_channels = extract_num_channels(input_image)
+    channel_dim = 1
+    momentum = 1e-1
+    configs = (
+        ((in_channels, in_channels), True, ("block1", "block_2")),
+        ((in_channels, in_channels), True, None),
+        ((in_channels, in_channels), False, ("block1", "block_2")),
+        ((in_channels, in_channels), False, None),
+    )
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:233]: We should test two different number of `in_channels` to be sure. Maybe something like `input_image1 = input_image` and `input_image2 = batch_up_image(input_image)`.

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 13]
```diff
@@ -120,3 +167,217 @@ def test_get_activation_module(subtests):
...
+            )
+
+        with subtests.test("out_channels"):
+            assert block.out_channels == sum(branch_in_channels)
+
+        with subtests.test("channel_dim"):
+            assert block.channel_dim == channel_dim
+
+        with subtests.test("forward"):
+            inputs = (input_image,) * 2
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:258]: With above suggestion
```suggestion
            inputs = (input_image1, input_image2)
```

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 14]
```diff
@@ -120,3 +167,217 @@ def test_get_activation_module(subtests):
...
+def test_UlyanovEtAl2016JoinBlock_num_channels_names_mismatch():
+    in_channels = 3
+    with pytest.raises(RuntimeError):
+        modules.UlyanovEtAl2016JoinBlock((in_channels, in_channels), names=("block1",))
+
+
+def test_UlyanovEtAl2016BranchBlock(subtests, input_image):
+    deep_branch = nn.Conv2d(3, 3, 1)
+    shallow_branch = nn.Conv2d(3, 3, 1)
+    momentum = 1e-1
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:287]: You should get this from `get_norm_module()`

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 15]
```diff
@@ -222,16 +222,21 @@ def test_UlyanovEtAl2016ConvSequence(subtests):
 
 
 def test_UlyanovEtAl2016JoinBlock(subtests, input_image):
-    in_channels = extract_num_channels(input_image)
+    input_image1 = input_image
+    input_image2 = torch.cat((input_image, input_image), 1)
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:222]: You could use `batch_up_image`

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 16]
```diff
@@ -222,16 +222,21 @@ def test_UlyanovEtAl2016ConvSequence(subtests):
...
-        ((in_channels, in_channels), True, None),
-        ((in_channels, in_channels), False, ("block1", "block_2")),
-        ((in_channels, in_channels), False, None),
+        (True, ("block1", "block_2")),
+        (True, None),
+        (False, ("block1", "block_2")),
+        (False, None),
     )
-    for branch_in_channels, instance_norm, names in configs:
+    for instance_norm, names in configs:
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py]: `for instance_norm, names in itertools.product((True, False), (("block1", "block2"), None))`:

File: tests/ulyanov_et_al_2016/test_modules.py [hunk 17]
```diff
@@ -222,16 +222,21 @@ def test_UlyanovEtAl2016ConvSequence(subtests):
...
 def test_UlyanovEtAl2016JoinBlock(subtests, input_image):
-    in_channels = extract_num_channels(input_image)
+    input_image1 = input_image
+    input_image2 = torch.cat((input_image, input_image), 1)
+    branch_in_channels = (
+        extract_num_channels(input_image1),
+        extract_num_channels(input_image2),
+    )
     channel_dim = 1
     momentum = 1e-1
```
Comment (pmeier) [inline tests/ulyanov_et_al_2016/test_modules.py:228]: See change before to `momentum`

File: tests/ulyanov_et_al_2016/test_modules.py (hunk 15 above)
Comment (jbueltemeier) [reply tests/ulyanov_et_al_2016/test_modules.py:222]: But the `batch_size` should not change here. The difference here is that the pyramid increases the `num_channels` with each level and enters the join_block as `shallow_branch`. For this reason the num_channels must be increased and not the batch_size.

Comment (pmeier) [reply tests/ulyanov_et_al_2016/test_modules.py:222]: Oops, you are right about that.

**Pull request merged**